src/sys/unix/kqueue.rs

```diff
@@ -307,7 +307,9 @@ impl Events {
             }
 
             if e.flags & libc::EV_EOF != 0 {
-                event::kind_mut(&mut self.events[idx]).insert(UnixReady::hup());
+                if e.filter == libc::EVFILT_READ as Filter {
```

Comment:
I think the if statement should be removed and hup should always be set if `EV_EOF` is set. From the FreeBSD kqueue manual [[1]]:

> `EVFILT_READ`:
> If the read direction of the socket has shutdown, then the filter also sets `EV_EOF` in `flags`, and returns the socket error (if any) in `fflags`.
>
> `EVFILT_WRITE`:
>The filter will set EV_EOF when the reader disconnects [...]

The way I read that is if either you can't read anymore (read direction was shutdown), or you can't write anymore (reader was disconnected) it sets the `EV_EOF` flag. The best way we can relay that information currently is via hup readiness.

[1]: https://www.freebsd.org/cgi/man.cgi?query=kqueue&manpath=FreeBSD+12.0-RELEASE+and+Ports